Classes/Core Modules/RPJSRequest.m

```diff
@@ -12,34 +12,114 @@
+        queue.qualityOfService = NSQualityOfServiceUserInitiated;
     });
-    return requestManager;
+    return queue;
+}
+
++ (AFHTTPRequestOperation *)operationWithURLString:(NSString *)urlString httpMethod:(NSString *)httpMethod headers:(NSDictionary *)headers parameters:(NSDictionary *)parameters successCallback:(JSValue *)successCallback errorCallback:(JSValue *)errorCallback {
+    if (!urlString || urlString.length == 0 || !httpMethod || httpMethod.length == 0) {
+        return nil;
+    }
+
+    AFJSONRequestSerializer *requestSerializer = [[AFJSONRequestSerializer alloc] init];
+    NSError *requestConstructionError;
+    NSMutableURLRequest *request = [requestSerializer requestWithMethod:httpMethod URLString:urlString parameters:parameters error:&requestConstructionError];
+    if (requestConstructionError) {
+        return nil;
+    }
+
+    NSMutableDictionary *mutableHeaders = [headers mutableCopy] ?: [NSMutableDictionary dictionary];
+    if (!mutableHeaders || mutableHeaders.count == 0) {
+        mutableHeaders = [[self defaultHeaders] mutableCopy];
+    }
+    else {
+        // "Extend" user headers dictionary with defaults
+        NSDictionary *defaultHeaders = [self defaultHeaders];
+        for (NSString *headerName in defaultHeaders) {
+            NSString *headerValue = defaultHeaders[headerName];
+            if (!mutableHeaders[headerName]) {
+                mutableHeaders[headerName] = headerValue;
+            }
+        }
+    }
+
+    for (NSString *headerName in mutableHeaders) {
+        NSString *header = mutableHeaders[headerName];
+        [request setValue:header forHTTPHeaderField:headerName];
+    }
+
+    AFHTTPRequestOperation *operation = [[AFHTTPRequestOperation alloc] initWithRequest:request];
+    operation.responseSerializer = [AFCompoundResponseSerializer compoundSerializerWithResponseSerializers:@[ [[AFHTTPResponseSerializer alloc] init], [[AFJSONResponseSerializer alloc] init] ]];
+    [operation setCompletionBlockWithSuccess:^(AFHTTPRequestOperation *completedOperation, id responseObject) {
+        NSString *responseString = [[NSString alloc] initWithData:responseObject encoding:NSUTF8StringEncoding];
+        if (responseString) {
+            [successCallback callWithArguments:@[ responseString ]];
+        }
+    } failure:^(AFHTTPRequestOperation *failedOperation, NSError *error) {
+        [errorCallback callWithArguments:@[ [[error userInfo] description] ?: @"" ]];
+    }];
+
+    return operation;
+}
+
++ (NSString *)urlStringFromURLValue:(JSValue *)urlValue {
+    NSString *urlString;
+
+    if ([urlValue isString]) {
+        urlString = [urlValue toString];
+    }
+    else if ([urlValue isObject]) {
+        NSDictionary *options = [urlValue toDictionary];
+        urlString = options[@"url"];
+    }
+
+    return urlString;
+}
+
++ (NSDictionary *)headersFromURLValue:(JSValue *)urlValue {
+    NSDictionary *headers;
```

Comment:
If we initialize this to an empty dictionary we can safely remove some of the safety checks above. Should lead to much simpler logic with the 2 changes.